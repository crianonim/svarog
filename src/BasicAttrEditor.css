.attr-wrapper > * {
    margin:auto 5px;
   }
   .attr-pair > * {
    margin:auto 2px;
   }
   