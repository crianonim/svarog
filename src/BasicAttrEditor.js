import React, { useState } from "react";
import './BasicAttrEditor.css';
import {attrsData} from './lib/helper.js';



const BasicAttrEditor = props => {
  const [controls, setControls] = useState(Object.entries(props.attrs));
  const validAttrs = Object.entries(attrsData)
   .filter( ([_,attr])=>!attr.el || attr.el.includes(props.element))
   .map(([key,_])=>key);
  const unUsedAttrs = validAttrs.filter(key =>
    controls.every(control => control[0] !== key)
  );
  const [addAttr, setAddAttr] = useState(unUsedAttrs[0]);
  // console.log(props.element, unUsedAttrs);
  return (
    <div className="flex-row flex-wrap attr-wrapper">
      <span>{props.element}</span>
      {props.edited?
      <>
      <select
      value={addAttr}
      onChange={e => {
        console.log(e.target.value);
        setAddAttr(e.target.value);
      }}
      >
        {unUsedAttrs.map(key => (
          <option key={key}>{key}</option>
          ))}
      </select>
      

      <button
        onClick={() => {
          console.log({ addAttr });
          const newControls = [...controls, [addAttr, attrsData[addAttr].def]];
          setControls(newControls);
          setAddAttr(unUsedAttrs.filter(key => key !== addAttr)[0]);
          props.changed(Object.fromEntries(newControls));
        }}
      >
        +
      </button>
      </>
      :null}
      {controls.map(([key, value], i) => {
        return (
          <span key={key} className="attr-pair">
            
            <span className="key-name">{key}</span>
            <input 
              onChange={e => {
                controls[i][1] = e.target.value;
                setControls(controls.slice());
                props.changed(Object.fromEntries(controls));
            }}
            style={{ width: (value+"").length / 2 + "rem" }}
              value={value}
            />
            {attrsData[key].type==="color"?(<span className="color-box" style={{color:value}}>&#x2588;</span>):null}
            {props.edited?(<button onClick={()=>{
              const changedAttrs=removeAttributeFromShape(controls,key)
              props.changed(Object.fromEntries(changedAttrs))
              setControls(changedAttrs);
              }
            }>x</button>):null}
          </span>

        );
      })}
    </div>
  );
};

const removeAttributeFromShape=(attributes,attributeName)=>{
  return attributes.filter( ([key,_])=>key!==attributeName );
}

export default BasicAttrEditor;
