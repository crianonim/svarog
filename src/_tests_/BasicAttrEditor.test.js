import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import BasicAttrEditor from "../components/BasicAttrEditor";

it("matches a snapshot",()=>{
    const wrapper=shallow(<BasicAttrEditor element="circle" attrs={{"x":10,"y":20,"r":30}}/>) 
    expect(toJson(wrapper)).toMatchSnapshot();
});
it("changes is fired when add property is clicked",()=>{
    const changed=jest.fn();
    const wrapper=shallow(<BasicAttrEditor changed={changed} element="circle" attrs={{"x":10,"y":20,"r":30}}/>) 
    wrapper.find("button.button.is-small.is-primary").simulate('click');
    expect(changed).toHaveBeenCalledTimes(1);
})
it("changes is called with new set attributes with new selected one",()=>{
    const changed=jest.fn();
    const wrapper=shallow(<BasicAttrEditor changed={changed} element="circle" attrs={{"x":10,"y":20,"r":30}}/>) 
    wrapper.find("button.button.is-small.is-primary").simulate('click');
    expect(changed).toBeCalledWith({ x: 10, y: 20, r: 30, fill: 'pink' });
})
it("changes is called without deleted attr",()=>{
    const changed=jest.fn();
    const wrapper=shallow(<BasicAttrEditor changed={changed} element="circle" attrs={{"x":10,"y":20,"r":30}}/>) 
    wrapper.find("button.button.is-small.is-danger").first().simulate('click');
    expect(changed).toBeCalledWith({y: 20, r: 30 });
})
it("has 3 attributes and 3 inputs",()=>{
    const wrapper=shallow(<BasicAttrEditor element="circle" attrs={{"x":10,"y":20,"r":30}}/>) 
    const attrs=wrapper.find('.key-name')
    expect(attrs.length).toBe(3)
    const inputs=wrapper.find('input');
    expect(inputs.length).toBe(3);
})
it("updates attrs when value in input is changed",()=>{
    const changed=jest.fn();
    const wrapper=shallow(<BasicAttrEditor changed={changed} element="circle" attrs={{"x":10,"y":20,"r":30}}/>) 
    const input=wrapper.find('input').first();
    input.simulate('change',{target:{value:66}});
    expect(changed).toBeCalledWith({ x: 66, y: 20, r: 30 });

})