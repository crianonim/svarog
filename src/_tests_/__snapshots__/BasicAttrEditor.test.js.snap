// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches a snapshot 1`] = `
<div
  className="flex-row flex-wrap  block"
>
  <span>
    Add property: Â 
  </span>
  <span
    className="field has-addons"
  >
    <span
      className="control"
    >
      <span
        className="select   is-small"
      >
        <select
          onChange={[Function]}
          value="fill"
        >
          <option
            key="fill"
          >
            fill
          </option>
          <option
            key="stroke"
          >
            stroke
          </option>
          <option
            key="strokeWidth"
          >
            strokeWidth
          </option>
          <option
            key="cx"
          >
            cx
          </option>
          <option
            key="cy"
          >
            cy
          </option>
        </select>
      </span>
    </span>
    <span
      className="control"
    >
      <button
        className="button is-small is-primary"
        onClick={[Function]}
      >
        +
      </button>
    </span>
  </span>
  <span
    className="flex-row has-hmargin-med"
    key="x"
  >
    <span
      className="key-name"
    >
      x
      :  
    </span>
    <span
      className="field has-addons"
    >
      <span
        className="control"
      >
        <input
          className="input is-small    "
          onChange={[Function]}
          style={
            Object {
              "width": "3rem",
            }
          }
          value={10}
        />
      </span>
      <span
        className="control"
      >
        <button
          className="button is-small is-danger"
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f1f8",
                  "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
                ],
                "iconName": "trash",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </button>
      </span>
    </span>
  </span>
  <span
    className="flex-row has-hmargin-med"
    key="y"
  >
    <span
      className="key-name"
    >
      y
      :  
    </span>
    <span
      className="field has-addons"
    >
      <span
        className="control"
      >
        <input
          className="input is-small    "
          onChange={[Function]}
          style={
            Object {
              "width": "3rem",
            }
          }
          value={20}
        />
      </span>
      <span
        className="control"
      >
        <button
          className="button is-small is-danger"
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f1f8",
                  "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
                ],
                "iconName": "trash",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </button>
      </span>
    </span>
  </span>
  <span
    className="flex-row has-hmargin-med"
    key="r"
  >
    <span
      className="key-name"
    >
      r
      :  
    </span>
    <span
      className="field has-addons"
    >
      <span
        className="control"
      >
        <input
          className="input is-small    "
          onChange={[Function]}
          style={
            Object {
              "width": "3rem",
            }
          }
          value={30}
        />
      </span>
      <span
        className="control"
      >
        <button
          className="button is-small is-danger"
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f1f8",
                  "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
                ],
                "iconName": "trash",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </button>
      </span>
    </span>
  </span>
</div>
`;
